plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.password.ldap;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.UncheckedExecutionException;
+import io.airlift.log.Logger;
+import io.trino.plugin.base.ldap.LdapClient;
+import io.trino.plugin.base.ldap.LdapQuery;
+import io.trino.spi.classloader.ThreadContextClassLoader;
+import io.trino.spi.security.AccessDeniedException;
+import io.trino.spi.security.GroupProvider;
+
+import javax.inject.Inject;
+import javax.naming.NamingException;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.trino.collect.cache.SafeCaches.buildNonEvictableCacheWithWeakInvalidateAll;
+import static io.trino.plugin.password.ldap.LdapAuthenticator.containsSpecialCharacters;
+import static io.trino.plugin.password.ldap.LdapAuthenticator.replaceUser;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class LdapGroupProvider
+        implements GroupProvider
+{
+    private static final Logger log = Logger.get(LdapGroupProvider.class);
+    private static final Pattern EXTRACT_GROUP_PATTERN = Pattern.compile("(?i)cn=([^,]*),(.*)$");
```

Comment:
Instead of using PatternMatcher can we try something like this 
```    private String getRelativeDistinguishedName(String distinguishedName)
            throws NamingException {
        LdapName ldn = new LdapName(distinguishedName);
        List<Rdn> rdns = ldn.getRdns();
        if (rdns.isEmpty()) {
            throw new NamingException("DN is empty");
        }
        Rdn rdn = rdns.get(rdns.size()-1);
        if (rdn.getType().equalsIgnoreCase(groupNameAttr)) {
            String groupName = (String)rdn.getValue();
            return groupName;
        }
        throw new NamingException("Unable to find RDN: The DN " +
                distinguishedName + " is malformed.");
    }

Reply:
sure, modifying ...